Qubes packages (Its alive!)

CertainLach opened the pull request:
## Description of changes

You can initialize qubes database files using `qubes-create` from `qubes-core-admin`, and then you need to install some virtual machines - this can be done by running qvm-backup on QubesOS system, and qvm-backup-restore in NixOS.

After that, basic functionality will work.

Be careful with netvms, as dom0 updates will not work with them. Some loosening up of qubes security needs to be implemented, at least having "qvm-dom0-network-via-netvm" script (Which was present in QubesOS before, and now needs to be written from scratch?) would be a good start.

At this point, you're welcome for testing, but be aware that it is not finished yet.

![image](https://github.com/user-attachments/assets/2e008447-36cb-447d-9f5c-8fa007195a47)

Cc: @SigmaSquadron

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SigmaSquadron commented on pkgs/development/python-modules/qubes-core-qrexec/default.nix:
```diff
@@ -0,0 +1,140 @@
+  '';
+
+  passthru = {
+    # FIXME: Should be split into separate packages?
```
Yes. Splitting this will most certainly reduce the complexity of the derivation.

CertainLach replied:
Current approach is okayish after refactoring.

SigmaSquadron commented on pkgs/development/python-modules/qubes-core-qrexec/default.nix:
```diff
@@ -0,0 +1,140 @@
+        --replace-fail "/var/run/qubes/qrexec-agent-fdpass" "${agent}/bin/qrexec-agent-fdpass" \
+        --replace-fail "/var/run/qubes/qrexec-agent" "${agent}/bin/qrexec-agent" \
+        --replace-fail "/usr/lib/qubes/qubes-rpc-multiplexer" "${agent}/bin/qubes-rpc-multiplexer" \
+        --replace-fail "/usr/bin/qrexec-policy-exec" "/run/current-system/sw/bin/qrexec-policy-exec"
```
Definitely not a good idea to use `current-system`. Could we just use `$out` here?

CertainLach replied:
This is util package, and qrexec-policy-exec is located in python package, which depends on this one, thus forming a recursive dependency. I have some ideas on how to fix that.

CertainLach replied:
I don't think there is a good solution here, those packages are coupled too tightly, and we might want to depend on system environment instead.

We have too many runtime dependencies to resolve here, I think it all needs to be built as a part of one big derivation.

```mermaid
graph LR
  daemon --> util
  agent --> util
  python --> util
  
  util -- qrexec-policy-exec --> python
  util -- qrexec-agent-fdpass, qrexec-agent, qubes-rpc-multiplexer --> agent
  python -- qrexec-client-vm, qubes-rpc-multiplexer --> agent
  python -- qrexec-client --> daemon
```

CertainLach replied:
Cycles are resolved

SigmaSquadron commented on pkgs/development/python-modules/qubes-core-qrexec/default.nix:
```diff
@@ -0,0 +1,140 @@
+  installTargets = [ "install-base" "install-dom0" "install-vm" ];
+
+  postFixup = ''
+    mv $out/${python.sitePackages}/usr/bin $out/bin
```
This should be before `fixupPhase`.

SigmaSquadron commented on pkgs/top-level/python-packages.nix:
```diff
@@ -13199,6 +13199,17 @@ self: super: with self; {
+
+  qubes-core-qubesdb = toPythonModule (pkgs.qubes-core-qubesdb.override {
+    inherit python;
+    withPython = true;
```
Might be better to make the main derivation have `withPython ? true,` and drop the override.

CertainLach replied:
Well, we still need to build this package using specific python version, thus override is inevitable

SigmaSquadron commented on pkgs/development/python-modules/qubes-core-qrexec/default.nix:
```diff
@@ -0,0 +1,140 @@
+    gtk3
+  ];
+
+  # Should also be packaged as vm-only?
```
A split output package sounds good here IMO.

CertainLach replied:
I think current approach with passthru is okayish, as those packages are tightly bound.
Proper splitting will require a qubes package set, which I want to avoid.

SigmaSquadron commented on pkgs/development/python-modules/qubes-core-qrexec/default.nix:
```diff
@@ -0,0 +1,140 @@
+      mv $out/usr/* $out/
+      rm -rf $out/usr
+      # Scripts: qubes-rpc-multiplexer
+      cp ../lib/* $out/bin/
```
```suggestion
      cp ../lib/* $out/libexec/
```
Very optional. If it breaks something, revert to `$out/bin`. Having executable libraries on libexec would make this more cohesive with the `xen` package.

CertainLach replied:
It is installed to bit, it is just located in lib in source code.
I want however to split some of sbin binaries to libexec

SigmaSquadron commented on nixos/modules/virtualisation/qubes.nix:
```diff
@@ -0,0 +1,34 @@
+  inherit (lib) mkEnableOption mkIf;
+  cfg = config.virtualisation.qubes;
+in {
+  # FIXME: Should it be qubes-dom0 or smth?
```
Let's mirror the Xen module. dom0 stuff is at the top-level (`virtualisation.qubes.*`) and domU stuff is under `virtualisation.qubes.guest.*`.

The file itself should be `qubes-dom0.nix` though.

SigmaSquadron reviewed (commented):
This is wonderful!

CertainLach commented on pkgs/by-name/qu/qubes-core-qubesdb/package.nix:
```diff
@@ -0,0 +1,64 @@
+    hash = "sha256-vPv74tBD7elYNqpgKLFKAanMH8D18OdDj0xhmw8aWwM=";
+  };
+
+  buildPhase = ''
+    make -C daemon $makeFlags
+    make -C client $makeFlags
+  '' + optionalString withPython "make -C python $makeFlags";
```
"make -C dir" means run "make" in "dir" directory, it is not the same as "make target".

This package can probably be split into multiple, with source roots "daemon"/"client"/"python"

SigmaSquadron replied:
Oh, that's right. If a package can be split, it probably should be.

CertainLach commented:
Pushed other batch of qubes fixes, now you can actually somehow manage VMs using qubes-manager, given that you have installed templates somehow (E.g you can copy VMs from other qubes installation). Standard qvm-template doesn't work, because it uses RPM to install such packages.

@SigmaSquadron xen needs python module, do you think it is possible to split it from xen package?

At least separate output will do, but ideally it needs to be built fully independently, since xen is a large package.
Also, another dependency cycle has been formed, this time it can't be resolved as with qrexec.

CertainLach commented on nixos/modules/virtualisation/qubes.nix:
```diff
@@ -0,0 +1,52 @@
+      # FIXME: qrexec-policy-exec system path is hardcoded in qubes-core-qrexec.util
+      pythonPackages.qubes-core-qrexec
+      # FIXME: qubes-core-admin-client system path is hardcoded in qubes-core-admin
+      pythonPackages.qubes-core-admin-client
```
Those packages (qubes-core-admin, qubes-core-admin-client) can't be merged, thus we need to deal with circular dependency somehow.
We may use PATH and add qubes-core-admin-client to the PATH of qubes-core-admin, but I don't think that will do.

CertainLach replied:
Cycles are resolved

SigmaSquadron commented:
> @SigmaSquadron xen needs python module, do you think it is possible to split it from xen package?
> 
> At least separate output will do, but ideally it needs to be built fully independently, since xen is a large package. Also, another dependency cycle has been formed, this time it can't be resolved as with qrexec.

I'm all for splitting the Python stuff from the main package, but I have no idea how to do that or how circular dependencies are even supposed to work in the first place.



> Forgot to cherry-pick some changes from my workstation nixpkgs. Sorry for the dirty git history, I will rebase and clean it up after more polishing.

Feel free to dump everything in the draft PR. We can always beat the git history into a shape later.

CertainLach commented:
> how circular dependencies are even supposed to work in the first place.

Well, they don't, derivation circular dependencies needs to be resolved somehow. Some can be merged together, some can be updated to be loosely coupled (E.g use PATH or specific runtime environment variables), but they need to be fixed, derivation can't depend on itself, because derivation hash = hash of its inputs.

Right now, both of discovered dependency cycles are resolved using absolute system installation paths, and I'm not sure of a proper solution.

SigmaSquadron commented:
Some more information for the module:

Qubes' [boot parameters](https://github.com/QubesOS/qubes-vmm-xen/blob/0f8d32f8e919c07d608947301100229a3cf7a2af/xen.spec.in#L719-L726):
```nix
virtualisation.xen.bootParams = [ "smt=off" "gnttab_max_frames=2048" "gnttab_max_maptrack_frames=4096" ];
```

Use the C-based XenStore (as Qubes doesn't use `oxenstored` due to the lack of security support for untrusted driver domains):

```nix
virtualisation.xen.store.path = "${pkgs.qubes-vmm-xen}/bin/xenstored";
```

CertainLach commented:
> Also, any systemd services brought in by the module need to have their Install sections duplicated in Nix code. (notably, wantedBy.)

Yeah, of course, this is just a rough draft of how it should look, nothing work in this nixos module yet.

BTW, qubesd has non-critical crash on service stop when using python 3.11 due to new python features being used (https://github.com/QubesOS/qubes-core-admin/commit/23be00ccf4c8c39fdaef908170aa37f2e3c3790a), and I can't upgrade it to 3.12 because of meson regression in Xen, do you have idea on how to fix that? https://github.com/NixOS/nixpkgs/issues/253751
Should have specified that this problem only affects Xen, qubes packages are building with python 3.12 just fine by itself.

SigmaSquadron commented:
> and I can't upgrade it to 3.12 because of meson regression in Xen, do you have idea on how to fix that? #253751 Should have specified that this problem only affects Xen, qubes packages are building with python 3.12 just fine by itself.

The hacky solution is to `python = (python312Packages.python.withPackages (p: [ p.setuptools ]));` in Xen.

CertainLach commented:
Thrown even more broken code into the pile, now it is able to generate menu entries for VMs, and recognize PCI device names.

Nixos module is still broken, but now it might start most of the required things by itself.

CertainLach commented on nixos/modules/virtualisation/qubes-dom0.nix:
```diff
@@ -0,0 +1,117 @@
+      pkgs.qubes-core-qubesdb
+
+      # Those are not technically optional, but might be disabled for non-gui nixos?
+      # qvm-appmenus - expected in $PATH by some tools, TODO: patch/wrap packages instead?
```
I think it makes sense to have all qvm-X commands in PATH by default. Using it will also allow to fix some of dependency cycles.

It would be nice if the system worked without them, but I don't think it is possible without potentially huge QubesOS refactoring.

CertainLach replied:
There is currently no dependency cycles, but some parts of the stack assume qvm-* commands are in PATH.
I don't think it is worth rewrapping everything, Its fine as long as there is no dependencies with hardcoded /run/current-system paths

CertainLach commented:
> The hacky solution is to python = (python312Packages.python.withPackages (p: [ p.setuptools ])); in Xen.

This hacky solution sounds good enough for now, I don't know how many of the bugs are caused by python version older than upstream wants.

CertainLach commented on pkgs/applications/qubes/qubes-core-vchan-xen/default.nix:
```diff
@@ -24,13 +24,17 @@ stdenv.mkDerivation rec {
     make install DESTDIR=$out PREFIX=/
   '';
 
+  # FIXME: Should only be enabled on xen 4.18+. It is also possible to make this detection in makefile,
```
I'll hardcode some of similar version check results, I don't think it is worth to support those.

Still it would be better if upstream performed those checks using pkg-config, and not using `ls /usr/includes/xen/xs.h` etc

CertainLach commented:
Latest changes make it run in KDE Wayland session.
![image](https://github.com/user-attachments/assets/2e008447-36cb-447d-9f5c-8fa007195a47)

There are still many bugs, but basic functionality works perfectly fine.

CertainLach commented on nixos/modules/virtualisation/qubes-dom0.nix:
```diff
@@ -0,0 +1,272 @@
+    }) list
+  );
+  # mkMergedEtc can only perform shallow copy, but deep is needed here.
+  mkMergedEtcByCopy = path: list: {
```
I think symlinkJoin will do just fine here.

marmarek commented on nixos/modules/virtualisation/qubes-dom0.nix:
```diff
@@ -0,0 +1,348 @@
+    !(builtins.elem flag cfg.optOutRecommendedConfiguration);
+
+in {
+  # Bug:
+  #   Loopback vchan connection not supported
+  #   FATAL: vchan initialization failed
+  # Solution: Xenstore has no set domid
+  #   It can be set manually using
+  #   `run0 xenstore-write "/domid" 0`
+  #   However it should always be set by default by something,
+  #   and I am not sure how/where.
+  # FIXME: Is this still relevant? I don't remember where I have encountered this error,
+  # but I no longer have this problem.
+
+  # Bug:
+  #   Not enough memory during start of VM
+  # Solution:
+  #   Qubes memory manager overrides max memory of dom0 with
+  #   total memory available, you need to set memory limit for it.
+  #   I have no idea what writes this memory limit in QubesOS, but
+  #   it can be set manually:
+  #   `run0 xenstore-write /local/domain/0/memory/static-max 100000000`
+  #   Note that the standard `xl mem-set` doesn't work for that, as it
+  #   uses /vm/uuid/memory field instead.
+  # FIXME: Permanent solution needed.
```
Both of those looks like missing `xen-init-dom0` call (normally a systemd service of this name, calling `/usr/libexec/xen/bin/xen-init-dom0`) - part of xen package

SigmaSquadron replied:
The Xen module is enabled [here](https://github.com/NixOS/nixpkgs/pull/341215/files#diff-9abfaa3b06166d20084f02a1d2bda37f9e492a2c7b017b72a37d017c0141171cR298), and that in turn enables the `xen-init-dom0` systemd service [here](https://github.com/NixOS/nixpkgs/blob/c6b2d0f9ade2f273e31b72c2acfef1cc1db752b6/nixos/modules/virtualisation/xen-dom0.nix#L832-L835).

Well, that's what should happen. In theory. It works with vanilla Xen.

CertainLach replied:
First problem is fixed even without domid key.
I believe it was caused by hostname being not dom0, thus causing qubes-core-admin-client to misbehave, because it can't find domain according to hostname (qubes expects adminvm to have name == dom0 with no ability to change it, and if dom0 hostname is not dom0 - then it can't detect that current domain is dom0)

This was fixed by changing default dom0 hostname to dom0, and by setting qubesdb /name key to dom0 (Which is used as an alternative method of hostname detection) for the cases where you can't use dom0 as hostname.

marmarek replied:
> qubes expects adminvm to have name == dom0 with no ability to change it

Right... I know it's stupid, but unfortunately it's hardcoded in too many places right now...

CertainLach replied:
qubesdb /name workaround works well for everything

CertainLach commented:
This branch is a little bit outdated, I'm debugging issues occured after nixos rebase.

For some reason, installation of StandaloneVM stopped working, qemu window appears, but it is always empty.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-xen-project-reviving-the-xen-project-hypervisor-on-nixpkgs/52768/7

CertainLach committed:
qubes-core-vchan-xen: 4.1.4 -> 4.2.4

CertainLach committed:
qubes-core-qubesdb: init at 4.2.6

CertainLach committed:
qubes-core-qrexec: init at 4.2.22

CertainLach committed:
qubes-core-admin: init at 4.3.6

CertainLach committed:
qubes-core-admin-client: init at 4.3.4

CertainLach committed:
qubes-manager: init at 4.2.9-1

CertainLach committed:
nixos/qubes: init module

CertainLach committed:
python3Packages.xen: init

CertainLach committed:
qubes-linux-utils: init at 4.3.4

CertainLach committed:
python3Packages.qubes-imgconverter: init

CertainLach committed:
python3Packages.qubes-desktop-linux-common: init at 4.2.12

CertainLach committed:
qubes-core-libvirt: init at 10.5.0-1

SigmaSquadron committed:
qubes-vmm-xen: init at 4.19.0-3

The first (and hopefully the last) custom Xen to be built based on the
generic Xen builder. It's just a standard build of Xen with a lot of
patches.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
qemu_qubes: init at 4.19.0-3

This package is required, as the Qubes Xen build is effectively a
xen-slim build.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

CertainLach committed:
qubes-seabios: init at 4.2.0

CertainLach committed:
qubes-vmm-linux-stubdom: init at 4.3.0-1

CertainLach committed:
plasma6: use xwayland module

CertainLach committed:
qubes-gui-common: init at 4.2.5

CertainLach committed:
qubes-app-linux-usb-proxy: init at 1.3.1

CertainLach committed:
qubes-desktop-linux-kde: init at 4.2.0

CertainLach committed:
qubes-core-admin-linux: init at 4.3.2

CertainLach committed:
qubes-core-agent-linux: init at 4.2.37

CertainLach committed:
qubes-artwork: init at 4.3.2-1

CertainLach committed:
usbguard: packages option

CertainLach committed:
pkgs.symlinkJoin: ability to remove prefix from tree

CertainLach commented:
Rebased&Cleaned up the whole history.

StandaloneVM installation is still broken, I believe this is caused by some upstream (qubes or nixos) changes, as I don't see any relevant changes between rebased and not rebased version of this PR.

CertainLach commented on pkgs/by-name/qu/qubes-core-qubesdb/0001-fix-cmd-name-parsing.patch:
```diff
@@ -0,0 +1,36 @@
++    // in QubesOS that works well: /bin/qubesdb-read => read, however in nixos, path before can contain a -:
++    // /nix/store/aaaaaa-qubesdb/qubesdb-read => qubesdb/qubesdb-read, which is not a valid qubesdb command.
++    // Fix it by first taking a filename (=removing all components ending in /) of a path.
++    cmd_argv0 = argv[0];
```
TODO: replace with basename() call, and submit this patch upstream

CertainLach commented on pkgs/by-name/qu/qubes-gui-daemon/0001-fix-make-shmoverride-initfirst.patch:
```diff
@@ -0,0 +1,25 @@
+ shmoverride.so: shmoverride.o ./list.o
+ 	$(CC) $(CFLAGS) $(extra_cflags) -shared -o shmoverride.so \
+-		shmoverride.o list.o -ldl -lxenctrl -lxengnttab -Wl,-Bsymbolic
++		shmoverride.o list.o -ldl -lxenctrl -lxengnttab -Wl,-Bsymbolic -Wl,-z,initfirst
```
Help of someone with dynamic knowledge is wanted here.

While -z initfirst helps on NixOS, it causes problems in upstream, and I still have no idea why LD_PRELOAD not implies initfirst, and if it is correct to use initfirst here.

https://github.com/QubesOS/qubes-gui-daemon/pull/148

CertainLach replied:
A temporary fix is merged upstream.

CertainLach commented on pkgs/development/python-modules/qubes-core-admin/0005-fix-data_percent-might-be-empty.patch:
```diff
@@ -0,0 +1,56 @@
+From 3d56e8b0f177f98914cecbe0d45afefa22c952ec Mon Sep 17 00:00:00 2001
+From: Yaroslav Bolyukin <iam@lach.pw>
+Date: Mon, 16 Sep 2024 18:54:21 +0200
+Subject: [PATCH 5/6] fix: data_percent might be empty
```
TODO: Submit this patch upstream.

It is possible to include necessary changes in nixos module, but those changes may break workflows of some users.

CertainLach replied:
This patch doesn't work well.
Resulting errors might be not good, but I think it can be solved with better documentation/error message clarificatios.

CertainLach commented on pkgs/development/python-modules/qubes-core-admin/0002-fix-broken-qvm-device-symlinks.patch:
```diff
@@ -0,0 +1,31 @@
+From dadc0d427eddaf49a6b213e7345c86e81a9c84e7 Mon Sep 17 00:00:00 2001
+From: Yaroslav Bolyukin <iam@lach.pw>
+Date: Mon, 16 Sep 2024 18:54:05 +0200
+Subject: [PATCH 2/6] fix: broken qvm-device symlinks
```
TODO: This and relevant qubes-core-admin-client needs to be submitted upstream

CertainLach commented on pkgs/development/python-modules/qubes-core-admin-client/0003-fix-do-not-crash-on-pool-with-no-volumes.patch:
```diff
@@ -0,0 +1,25 @@
+From 2448b386505af8540f21b0f81275f45776070f8c Mon Sep 17 00:00:00 2001
+From: Yaroslav Bolyukin <iam@lach.pw>
+Date: Sat, 14 Sep 2024 00:09:19 +0200
+Subject: [PATCH 3/3] fix: do not crash on pool with no volumes
```
TODO: Submit upstream

This is not an issue on QubesOS, because in qubes you have OOTB setup, and pools are never empty.

This is not the case on NixOS, on Nix you need to install distributions after.

CertainLach replied:
Upstream PR: https://github.com/QubesOS/qubes-core-admin-client/pull/311

CertainLach replied:
Merged upstream

CertainLach commented:
Figured out StandaloneVM - this is caused by changes that were not included in the previous version of this PR, but in my private nixpkgs fork
```nix
    ++ lib.lists.optional withSeaBIOS "--with-system-seabios=${seabios}/share/seabios"
```

This is incorrect, as `--with-system-seabios` accepts full path to seabios binary, after replacing path with the correct one StandaloneVM installation works again.
TODO: Full path needs to be overrideable too, as we want to use qubes-seabios here, and it has different binary names inside.

@SigmaSquadron PTAL

SigmaSquadron reviewed (commented):
The build failures can probably be fixed with a rebase once #345171, #345192 and #341429 are merged.

---

Result of `nixpkgs-review pr 341215` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>65 packages failed to build:</summary>
  <ul>
    <li>appvm</li>
    <li>collectd</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>docker-machine-kvm2</li>
    <li>gnome-boxes</li>
    <li>guestfs-tools</li>
    <li>htcondor</li>
    <li>libguestfs</li>
    <li>librenms</li>
    <li>libvirt</li>
    <li>libvirt-glib</li>
    <li>libvirt-glib.dev</li>
    <li>libvirt-glib.devdoc</li>
    <li>libvmi</li>
    <li>libvmi.dev</li>
    <li>libvmi.lib</li>
    <li>mgmt</li>
    <li>minikube</li>
    <li>multipass</li>
    <li>perl536Packages.SysVirt</li>
    <li>perl536Packages.SysVirt.devdoc</li>
    <li>perl538Packages.SysVirt</li>
    <li>perl538Packages.SysVirt.devdoc</li>
    <li>python311Packages.guestfs</li>
    <li>python311Packages.guestfs.dist</li>
    <li>python311Packages.libvirt</li>
    <li>python311Packages.libvirt.dist</li>
    <li>python312Packages.guestfs</li>
    <li>python312Packages.guestfs.dist</li>
    <li>python312Packages.libvirt</li>
    <li>python312Packages.libvirt.dist</li>
    <li>qemu_xen_4_17</li>
    <li>qemu_xen_4_17.debug</li>
    <li>qemu_xen_4_17.ga</li>
    <li>rubyPackages.ruby-libvirt (rubyPackages_3_3.ruby-libvirt)</li>
    <li>rubyPackages_3_1.ruby-libvirt</li>
    <li>rubyPackages_3_2.ruby-libvirt</li>
    <li>vagrant</li>
    <li>virt-manager</li>
    <li>virt-manager-qt</li>
    <li>virt-manager.dist</li>
    <li>virt-top</li>
    <li>virt-viewer</li>
    <li>xen (xenPackages.xen_4_19)</li>
    <li>xen.boot (xenPackages.xen_4_19.boot)</li>
    <li>xen.dev (xenPackages.xen_4_19.dev)</li>
    <li>xen.doc (xenPackages.xen_4_19.doc)</li>
    <li>xen.man (xenPackages.xen_4_19.man)</li>
    <li>xenPackages.xen_4_17</li>
    <li>xenPackages.xen_4_17-slim</li>
    <li>xenPackages.xen_4_17-slim.boot</li>
    <li>xenPackages.xen_4_17-slim.dev</li>
    <li>xenPackages.xen_4_17-slim.doc</li>
    <li>xenPackages.xen_4_17-slim.man</li>
    <li>xenPackages.xen_4_17.boot</li>
    <li>xenPackages.xen_4_17.dev</li>
    <li>xenPackages.xen_4_17.doc</li>
    <li>xenPackages.xen_4_17.man</li>
    <li>xenPackages.xen_4_18</li>
    <li>xenPackages.xen_4_18.boot</li>
    <li>xenPackages.xen_4_18.dev</li>
    <li>xenPackages.xen_4_18.doc</li>
    <li>xenPackages.xen_4_18.man</li>
  </ul>
</details>
<details>
  <summary>66 packages built:</summary>
  <ul>
    <li>python311Packages.qubes-app-linux-usb-proxy</li>
    <li>python311Packages.qubes-core-admin</li>
    <li>python311Packages.qubes-core-admin-client</li>
    <li>python311Packages.qubes-core-admin.man</li>
    <li>python311Packages.qubes-core-libvirt</li>
    <li>python311Packages.qubes-core-libvirt.dist</li>
    <li>python311Packages.qubes-core-qrexec</li>
    <li>python311Packages.qubes-core-qubesdb</li>
    <li>python311Packages.qubes-core-qubesdb.dist</li>
    <li>python311Packages.qubes-desktop-linux-common</li>
    <li>python311Packages.qubes-desktop-linux-common.dist</li>
    <li>python311Packages.qubes-imgconverter</li>
    <li>python311Packages.qubes-imgconverter.dist</li>
    <li>qubes-vmm-xen (python311Packages.qubes-vmm-pyxen)</li>
    <li>qubes-vmm-xen.boot (python311Packages.qubes-vmm-pyxen.boot)</li>
    <li>qubes-vmm-xen.dev (python311Packages.qubes-vmm-pyxen.dev)</li>
    <li>qubes-vmm-xen.doc (python311Packages.qubes-vmm-pyxen.doc)</li>
    <li>qubes-vmm-xen.man (python311Packages.qubes-vmm-pyxen.man)</li>
    <li>python312Packages.qubes-app-linux-usb-proxy</li>
    <li>python312Packages.qubes-core-admin</li>
    <li>python312Packages.qubes-core-admin-client</li>
    <li>python312Packages.qubes-core-admin.man</li>
    <li>python312Packages.qubes-core-libvirt</li>
    <li>python312Packages.qubes-core-libvirt.dist</li>
    <li>python312Packages.qubes-core-qrexec</li>
    <li>python312Packages.qubes-core-qubesdb</li>
    <li>python312Packages.qubes-core-qubesdb.dist</li>
    <li>python312Packages.qubes-desktop-linux-common</li>
    <li>python312Packages.qubes-desktop-linux-common.dist</li>
    <li>python312Packages.qubes-imgconverter</li>
    <li>python312Packages.qubes-imgconverter.dist</li>
    <li>qemu_qubes</li>
    <li>qemu_qubes.debug</li>
    <li>qemu_qubes.ga</li>
    <li>qemu_xen (qemu_xen_4_19)</li>
    <li>qemu_xen.debug (qemu_xen_4_19.debug)</li>
    <li>qemu_xen.ga (qemu_xen_4_19.ga)</li>
    <li>qemu_xen_4_18</li>
    <li>qemu_xen_4_18.debug</li>
    <li>qemu_xen_4_18.ga</li>
    <li>qubes-artwork</li>
    <li>qubes-core-admin-linux</li>
    <li>qubes-core-agent-linux</li>
    <li>qubes-core-agent-linux.man</li>
    <li>qubes-core-libvirt</li>
    <li>qubes-core-qubesdb</li>
    <li>qubes-core-vchan-xen</li>
    <li>qubes-desktop-linux-kde</li>
    <li>qubes-gui-common</li>
    <li>qubes-gui-daemon</li>
    <li>qubes-linux-utils</li>
    <li>qubes-manager</li>
    <li>qubes-manager.dist</li>
    <li>qubes-seabios</li>
    <li>qubes-vmm-stubdom-linux</li>
    <li>xen-guest-agent</li>
    <li>xen-slim (xenPackages.xen_4_19-slim)</li>
    <li>xen-slim.boot (xenPackages.xen_4_19-slim.boot)</li>
    <li>xen-slim.dev (xenPackages.xen_4_19-slim.dev)</li>
    <li>xen-slim.doc (xenPackages.xen_4_19-slim.doc)</li>
    <li>xen-slim.man (xenPackages.xen_4_19-slim.man)</li>
    <li>xenPackages.xen_4_18-slim</li>
    <li>xenPackages.xen_4_18-slim.boot</li>
    <li>xenPackages.xen_4_18-slim.dev</li>
    <li>xenPackages.xen_4_18-slim.doc</li>
    <li>xenPackages.xen_4_18-slim.man</li>
  </ul>
</details>

CertainLach committed:
fixup! pick seabios changes from https://github.com/NixOS/nixpkgs/pull/341429

CertainLach committed:
qubes-core-admin: drop lvm patch

CertainLach committed:
qubes-core-admin: add meta

CertainLach committed:
qubes-manager: add meta

CertainLach committed:
qubes-gui-daemon: less aggressive library loading fix

CertainLach committed:
qubes-gui-daemon: use python3Packages

CertainLach committed:
qubes-gui-daemon: nixfmt

CertainLach committed:
qubes-core-admin: fixup missing patch

ethnh commented:
god bless you seriously :bow:

CertainLach commented:
Another rebase, this time StandaloneVM display is broken again?
This time seabios path is passed correctly as I can see, is it broken due to different reasons, or something in seabios/xen build has been broken in the meantime?

CertainLach commented:
I have made some progress on QubesOS builders and building stubdom for NixOS-Qubes, however I believe it would be much easier to handle those tasks after base QubesOS packages are already merged in NixOS, even if they have suboptimal implementation (Looking at stubdom, which is extracted from QubesOS RPMs) sometimes.

Another way to simplify management of this package set would be extracting them to separate flake... Thoughts?

For now, I have updated all of packages to latest versions, rebased to latest master, and populated meta attributes as much as possible, to mark it ready for review.

Please ignore formatting deficiencies for now (Some of patches not present here might break on reformatting, and I would like to avoid too much rebase conflicts), there is enough of crimes against humanity here without that.

CertainLach commented:
StandaloneVM seems to be fixed again, but I would expect that it might break again due to unknown reasons, I believe this might be fixed after unification of package versions between dom0 and stubdom, which will only happen in the future, as building our own stubdom is too complicated for now.

evq commented:
this is awesome work @CertainLach!

I've been working on Qubes packaging as well, coming at it from the angle of just creating a NixOS template vm that can be used under vanilla Qubes. I love the idea of using NixOS in dom0 as well.

as I'm fairly new to Nix I packaged it as a standalone flake to get feedback on how to best upstream it:
https://github.com/evq/qubes-nixos-template

if what I've created is useful it'd be great to collaborate on merging the two.

hustlerone commented:
Do all of the qubes packages integrate well with nixos? I have used Qubes and there has been some jank because of the way it's got to do things

CertainLach commented:
As it became too painful to maintain this patchset as a part of nixos, I decided to split it into different git repo for now.
https://github.com/CertainLach/nixos-qubes

This repo follows the same naming as NixOS, so patches from it should be also appliable to NixOS, but for quicker iteration it exposes itself as an overlay+flake.
It doesn't work right now, after I updated all the qubes packages to their latest versions (namely, Xen fails to build), but now it would be easier to fix than doing this inside of nixpkgs tree.

After all the mandatory functionality will become useable for normal user without too much tinkering (ideally, enabling qubes option should be enough) - all patches from this repo will be applied to NixOS in a new pull request.

CertainLach commented:
> Do all of the qubes packages integrate well with nixos? I have used Qubes and there has been some jank because of the way it's got to do things

As far as dom0 goes, I haven't encountered any problems. I even have prototype of update daemon, which delivers updates from nixos cache to dom0 using updatevm.

However, I haven't worked with NixOS as TemplateVM, and I think there might be some problems with store.

hustlerone commented:
Best of luck. I'll start messing with it once things start working

CertainLach commented:
![image](https://github.com/user-attachments/assets/93f8dd69-c6d3-449f-b091-522b60dcea54)

Qubes in [nixos-qubes](https://github.com/CertainLach/nixos-qubes) repo is working.

Quick-start guide:
```nix
# Add flake input
    nixos-qubes = {
      url = "github:CertainLach/nixos-qubes";
      inputs.nixpkgs.follows = "nixpkgs";
    };
# Add nixos module
    modules = [
                inputs.nixos-qubes.nixosModules.default
    ];
# Enable qubes
  virtualisation.qubes = {
    dom0.enable = true;

    # Current qubes module implementation is not as secure as QubesOS, this option is mandatory to clarify that.
    secure = false;
    
    # Insert your primary username here, qubes is single-user system. Only this user will be able to use qubes utilities.
    user = USERNAME;

    # All those options are opt-outs, I.e only keep options here that you want to be active
    # Configuration I provide here is supposed to keep your system running as-is for now, 
    # I.e you will have USB, internet access, and qubes module will not mess with your system configuration too much.
    optOutRecommendedConfiguration = [ 
      "disable-smt" # Do not disable SMT
      "hostname-dom0" # Use current machine hostname, do not reset it to dom0
      "dom0-restricted-usb" # Do not block USB devices in dom0
      "dedicated-sys-usb" # Allow USB devices in dom0
      "dedicated-sys-net" # Allow networking in dom0
    ];
  };
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-xen-project-reviving-the-xen-project-hypervisor-on-nixpkgs/52768/11

Sigmanificient commented:
This is fantastic!
